tailscale/tailscale
# util/linuxfw: add nftable support for linux router

### Summary:

This change implements `netfilterRunner` interface for nftables manipulation as a low level nftables runner. This change allows manipulation to netfilter rules via nftables. 

### Issue related

Updates: #391 

### Changes made

1. Added `nftableRunner` that implements `netfilterRunner` and `nftable` as a container of the `nftables.Table` abstraction.
2. Added envknob `TS_DEBUG_USE_NETLINK_NFTABLES` Which controls using nftables or iptables. 
3. Added nftables rule creaters via google/nftables package.
5. Added unit test for nftables, the tests are splitted in to two parts. The first part tests on the correctness of rule we insert against the expected binary representation. The second part tests if the rules are added and deleted by interface methods in a testing namespace.


### Thought Process

1. `nftablesRunner` contains a single connection for all nftables operation. This design was for testing with customized connections.
2. The second part of tests has to run with sudo privilege since we created a new network namespace for testing the rule operations.
3. All ts rules has `counter` expression for basic debug log purposes provided to user. If we want we can also add `log` expressions for logging in system
4. Since nftables doesn't have default table, we also created tailscale tables to contain the chains.
5. Since nftables doesn't have default chains, we have promoted our chain to a little bit higher priority than usual (-1 for filter type chains and -100 for NAT type chain) to be processed ahead of "normal fw rules" (basically the ones that follow convensions). **The NAT type chain has reached highest possible priority but filter type chain can be promoted more if we want to.**
6. Table cleaning was implemented for nftables rules.

### Expected Result

tailscaled should be manipulating rules using nftables when it's running with envknob `TS_DEBUG_USE_NETLINK_NFTABLES=true` set. We expect that the app works the exact same way as we use iptables, and also expect [these rules](https://paste.corp.ts.net/paste/857dc61d-282f-4c9b-b40a-bd94a2a2d6f4) to be present when we run `sudo nft list ruleset`. Notice that the rule matching `iifname "lo*"` has local machine address as saddr, and the rules have the bytes displayed in little endian in the file paste file)

As an example for IPv4 rules.

### How to test

On linux environment do the following:

1. Bring up tailscaled by running `sudo TS_DEBUG_USE_NETLINK_NFTABLES=true ./tool/go run ./cmd/tailscaled`
6. Start tailscale with a testing tailnet on corp.
8. Run `sudo nft list ruleset` to view the rules added.
9. Test the CGNAT rule and ChromeOSVMRange rule by creating packet from corresponding saddr range to a none tailscale network interface.
10. Shut down tailscale
11. Run tailscale as an exit node by running `sudo ./tool/go run ./cmd/tailscale up --advertise-exit-node`, the way of allowing forward and exit node can be found [here](https://tailscale.com/kb/1103/exit-nodes/?tab=linux).
12. Connect to exit node from another node on tailscale.
13. Check traffic on NAT table and ts-forward chain respectly. (About setting and checking ts marks and controlling outgoing packets)

## Note
If we want to build then run, the env var should come with run, not build. So basically something like 
```
go build -o tailscaled ./cmd/tailscaled/ && env TS_DEBUG_USE_NETLINK_NFTABLES=1 sudo -E ./tailscaled --statedir /tmp/ts/ --socket /tmp/ts/ts1.sock --verbose 99
```
